### ISSUE TITLE
How do "refresh commands" work in the pootle app?

### ISSUE THREAD

**Issue by JohnnyKing94** [translate/pootle#4895]
How does "refresh_scores" works? I ask this cause i made some tests and it seems that did nothing. I wonder the same for "refresh_stats" in the second test i made.

## 1^ Test
I had some projects, then i removed all of them and i checked the db if the all the strings were gone. I executed the command `pootle refresh_scores` and i got this this output. Basically everything seemed to be working
```
2016-07-14 10:04:13,315 INFO Loading custom settings from '/root/.pootle/pootle.conf'...
2016-07-14 10:04:14,734 INFO Using Python PO
```
But then i checked my profile score and the scores of other profiles and I realized nothing changed, i had the same score before of removing the projects. The only way to change it, it was changing the value of the `score` field in `accounts_user` table. I even flushed the redis db with the proper command `redis-cli -n 1 flushdb`. Nothing!

## 2^ Test
I created a project with few strings in it, i made some translations with some test accounts, then i re-uploaded the project with two strings removed, i executed `pootle update_stores --project=<project_name>` and the lines were removed correctly. Later i ran `pootle refresh_stats` and `pootle refresh_scores`, nothing were changed... i had no errors in the rq worker, no failed jobs... nothing... but nevertheless i kept having the same scores and the same stats. I checked in statistics pages, nothing changed there too.
Then there was even this translation feed on the project home `<User> reviewed <text_string> An hour ago`. I clicked on it and I got **Not found** cause the string was removed as planned, but still i ketp having this feed near the project.

I don't know what i'm doing wrong, maybe it's supposed to work like that, maybe i'm missing something or what, so pls explain to me...

**Comment by ta2-1** [comment]
hi @JohnnyKing94 

re test1:
`refresh_scores` handles data in `pootle_statistics_scorelog` table (it is linked with `pootle_store_unit` via `pootle_app_submission`). Can you check if you removed related data when you removed your projects?
`refresh_scores` was created to recalculate all scores after you change score settings [here](https://github.com/translate/pootle/blob/45b81e265332bfcf681ae174d72ff0cd0a81c4a0/pootle/settings/40-apps.conf#L138).

re test2:
When you run `update_stores` it triggers all refresh jobs automatically. You don't need to run anything else. 

But I see your problem and I think there is a bug. `Last activity` is not updated after units are marked as obsolete. And `refresh_stats` doesn't check if unit is obsolete when calculates last activity message. I'll check this.

**Comment by JohnnyKing94** [comment]
## test1
before the test, the `pootle_statistics_scorelog` is full of data, when i removed the project via pootle admin section (no hard ways via db), the data in the table were gone. So basically if i run the command of `refresh_scores`, **it must reset my count to 0.** Why doesn't do that?

## test2
If `update_stores` triggers all other commands, why i keep having old stats of removed strings and an old count for the scores?

## I made a third test
I manually changed the `score` field in `accounts_user` with projects first and then without. In the first case of using r`efresh_scores` command, **the field should have been reset to 0** as there were no data to calculate from `pootle_statistics_scorelog`. In the second case, **the command should have recounted the scorelogs and fixed the value**. Nothing of this happened!

**Comment by ta2-1** [comment]
Thanks @JohnnyKing94, there are 2 bugs you have reported about:
- https://github.com/translate/pootle/issues/4897
- Last activity is not updated after units are marked as obsolete. And refresh_stats doesn't check if unit is obsolete when calculates last activity message. I still need to check for this. 

Let me ask more re:
> why i keep having old stats of removed strings and an old count for the scores?

re stats: 
Are you talking about last activity only? I think there is a bug and we'll address it before 2.8.0 release.

re scores:
Did you expect user scores were changed after you removed (marked as obsolete) 2 units via `update_stores`?

**Comment by JohnnyKing94** [comment]
> Let me ask more re:
> 
>>     why i keep having old stats of removed strings and an old count for the scores?
> 
> re stats:
> Are you talking about last activity only? I think there is a bug and we'll address it before 2.8.0 release.

I was talking about the part you said before:
> When you run update_stores it triggers all refresh jobs automatically
So if i reupload a file (with some changed or removed strings) and then I run this command, it should:
- refresh the score (NOT WORKING)
- refresh the stats (NOT WORKING)
I keep seeing stats about removed/obsolete strings and i keep having the old score.

> Did you expect user scores were changed after you removed (marked as obsolete) 2 units via update_stores?
Not really, it would be great after the `update_stores` command, but it would also be great running the second command manually, I mean:
- First `update_stores` which remove, update strings correctly
- Second `refresh_score` and `refresh_stats` that should recount scores and stats again (**but NOT WORKING**)
If it could be done internally for the command `update_stores`, _it could be a good idea_... **but running 2 separately command, it makes no difference for me**
It's the same thing when you remove all the projects or some of them and it could be great if the score is recounted or reset to 0 in the process of deleting it (via admin panel), but it's not a big deal for me running the command in the console to refresh scores and stats

What i expect now is (IMPORTANT):
- `Last activity` must be updated when there are removed/changed/altered strings after the updating of a project
- The `refresh_scores` command should recount properly (the tests made seems that this command doesn't work at all both with `pootle_statistics_scorelog` table **populated** and **empty** cause i keep having the same score in both cases)
- The `refresh_stats` command must recount properly **WHEN** there are updated projects with variations within

Then if you would consider also these feedbacks, it would be great for members not using always the shell:
- Removing a single project via admin panel triggers the recounting process of the scores and the stats
- Removing all the projects via admin panel triggers the recounting process of the scores (the stats are wiped out from the table when you remove all the projects, the first test i made)
- Updating a project via `update_stores` command triggers the recounting process of the scores and the stats

**Comment by ta2-1** [comment]
@JohnnyKing94, you are right we didn't tested how stats and scores would be updated if you removed projects permanently via admin panel. That's definitely another issue and should be addressed.

Could you help me to understand another part of your issue better?
> So if i reupload a file (with some changed or removed strings) and then I run this command, it should:
- refresh the score (NOT WORKING)
- refresh the stats (NOT WORKING) I keep seeing stats about removed/obsolete strings and i keep having the old score.

What changes did you have in the file? Were those changes applied? I suspect if you reupload a file then it's revision wasn't updated and then no changes were applied.
I need minimal set of changes to reproduce this bug. It would be really helpful if you provide more detailed information here. I've just tried to upload .po file where I added translation for one untranslated unit:
- translation added;
- Last action:  <user> uploaded file An hour ago;
- <user> score updated;

And just to clarify:
User scores shouldn't be changed if some strings (unit) were marked as obsolete via `update_stores` command and no other changes were made. `refresh_stats` and `refresh_scores` should do nothing in this case.

**Comment by JohnnyKing94** [comment]
## Test about variations in translation files
> What changes did you have in the file? Were those changes applied after you upload the file? I suspect if you reupload a file then it's revision wasn't updated and then no changes were applied.
> I need minimal set of changes to reproduce this bug. It would be really helpful if you provide more detailed information here. I've just tried to upload .po file where I added translation for one untranslated unit:
> 
> - translation added;
> - Last action: uploaded file An hour ago;
> - score updated;

The only test i made is:

- uploading a .po file and creating the project
- then reuploading the file with one string removed via Poedit and a .pot template
- finally i updated the project, the string was correctly removed (2nd test) but the score did not change. same for the stats!

The strings weren't marked obsolete, they were permanently removed from the .po file using an updated .pot template.

## Tests about `refresh_scores` and `refresh_stats`
`refresh_scores` continues not to work correctly. Just do these tests below, but first check if you have `pootle_statistics_scorelog` **populated**:
- alter manually the field value `score` in `accounts_user` via **DB** with any values you wish
- run the `refresh_scores` command and check if your score is recounted correctly, If it is so, then it's my problem and i need help to figure it out, otherwise it's a bug

The other test you should is simply: just upload 3 or more projects, translate some strings for all the project uploaded, check your score, then **delete ONE** of them via admin panel and check again your score and your stats. If something doesn't seem correctly, run the `refresh_scores` command and the `refresh_stats`. Check if something changed, otherwise even this one it's a bug.

If you still have 2 project uploaded from the tests before, try to delete them all via admin panel and run the 2 commands to refresh, then re-check again both score and stats.

**Comment by JohnnyKing94** [comment]
I always forget to update this issue... i want to say that the issue was resolved updating to the new version 2.8.0b3 (it was not working on 2.8.0b2 if i remember)...

These are the following issues which are resolved
- **Test about variations in translation files** (both points)
- **Tests about refresh_scores and refresh_stats** (both two tests -- the manually editing and the removing of translation)

**The remaining issue for 2.8.0b3** is the permanently removing of projects and the missing recalculation of stats and scores (the one that also confirmed @ta2-1)

**Comment by unho** [comment]
Command is gone, thus the closing.
